nameko_sqlalchemy/database_session.py

```diff
@@ -42,3 +42,17 @@ def worker_teardown(self, worker_ctx):
 
 # backwards compat
 Session = DatabaseSession
+
+
+def run_query(session, query, attempts=2):
+    while attempts > 0:
+        attempts -= 1
+        try:
+            return query()
+        except exc.DBAPIError as exception:
+            if attempts > 0 and exception.connection_invalidated:
+                session.rollback()
+            else:
+                raise
+    else:  # pragma: no cover
```

Comment:
Is this needed?

Reply:
I know this looks bad but this is needed to keep the coverage on 100%. Apparently coverage expects to cover the case when the condition becomes false but that can never happen because either the result is returned or an exception is raised in the last iteration.

Reply:
I meant the `else:` clause actually, rather than the pragma.

Reply:
The `else` is there for the same reason. If it wasn't there with the pragma then the coverage would mark the line where the `while` loop starts as not covered.

Reply:
Oh I see. You're missing branch coverage. Pytest is pointing out that you've not tested the case where someone executes `run_query(..., attempts=0)`

This may go away if you implement the retry decorator, but a test that covered `run_query(..., attempts=0)` would be less messy than these extra bits of code.